### DISCUSSION TITLE
Using stoichometric matrix for graph generation

### DISCUSSION THREAD

**Discussion by formersbach** [biomass-dev/biomass#165] [Ideas]
hey @himoto !
Considering that with the changes in the kinetics branch we now have access to the stoichiometry matrix I intend to rework how the graph is generated. Everything is working fine, code is much more concise and simple.
However I noticed that reaction modulators aren't included in the stoichiometry matrix, therefore regulatory connections aren't transitioned to the graph. It's a similar problem I ran into while working on the sbml task.

Any comments on that or ideas to deal with that?
Thanks!

**Comment by himoto** [comment]
Hi @formersbach 
Can you  elaborate with a simple example? I would like to work on this.

**Reply by formersbach** [reply]
Using the following text file:
```
A + E <--> B
B --> C + E
C phosphorylates E --> pE
pE is dephosphorylated --> E
```
And the following (abbreviated) Text2ModelGraph class:
```python
def construct_adjacency_matrix(stoichometry_matrix) -> dict:
    """
    Matches species that interact with each other by reading python files as text and looking for equations.
    """
    adj_dim = stoichometry_matrix.shape[0]
    adj_matrix = np.zeros((adj_dim, adj_dim))
    for reaction in stoichometry_matrix.transpose():
        outnodes, innodes = [], []
        for i, species in enumerate(reaction):
            if species < 0:
                outnodes += [i]
            if species > 0:
                innodes += [i]
        adj_matrix[innodes, outnodes] = 1
    return adj_matrix

model = Text2Model("test.txt")
model.convert(overwrite=True)
print(construct_adjacency_matrix(model.stoichiometry_matrix.toarray()))
```
The output from this adj matrix is 0's for species C. In the final graph there should be a connection between C and E.
Hope the example helps!

**Reply by himoto** [reply]
Thank you for sharing necessary and sufficient example.
In line 3: `C phosphorylates E --> pE`,
Here C works as an enzyme and its amount affects flux, but this reaction will never affect the amount of C itself.
```
d[E]/dt = -v[3]
d[pE]/dt = + v[3]
# d[P]/dt does not depend on v[3]
```
Therefore in the resulting stoichiometry matrix, we cannot get any information on which enzyme (kinase, phosphatase, etc.) drives this reaction. I think it should not be included in the stoichiometry matrix but can store information about reaction modulators in some other place.
~~We can start from extending KineticInfo class by adding `modulators`:~~
(Aug. 6)
I am sorry but I didn't notice you've added `modifiers` to KineticInfo class.
```python
class KineticInfo(NamedTuple):
    reactants: tuple
    products: tuple
    modifiers: tuple
    rate: str
```
Do you think you can use `modifiers` for creating graph as well?

**Reply by himoto** [reply]
One additional question:
Currently graph visualization works in any models no matter whether the model is created via Text2Model or not.
However, the stoichiometry matrix is in `Text2Model` class and not available when the model is manually developed.
Will you drop support for manually developed models?

**Comment by formersbach** [comment]
> Therefore in the resulting stoichiometry matrix, we cannot get any information on which enzyme (kinase, phosphatase, etc.) drives this reaction. I think it should not be included in the stoichiometry matrix but can store information about reaction modulators in some other place.

I believe this depends on how the stoichiometry matrix (and the internal reactions) is set up. Take the classic michaelis menten system as an example: 
The overall reaction is simply S --> P. In isolation the stoichiometry matrix would just be [-1 1], but if you consider the reaction mechanism such that S + E <--> ES --> E + P the stoichiometry matrix would be
[-1 1 0
 -1 1 1
 1 -1 -1
 0  0 1]
 

> Do you think you can use modifiers for creating graph as well?

Yes that would be an alternative! However as it is now, modulator cannot specify direction of the modulation (inhibition or activation). If that's important we'll need to change modulator -> activator and inhibitor. 

>  Will you drop support for manually developed models?

Currently I'm only working on Text2Model, however as you mentioned in #150:
> Yes, currently this is only available for the Text2Model object. If you think the kinetics info is useful, I am planning to copy it in biomass.ModelObject class as well. Please note that we always require models to be created from Text2Model.

I believe it would be very useful, same for the stoichiometry matrix.
